### DISCUSSION TITLE
How out-of-the-box (noob-friendly) should the experience become?

### DISCUSSION THREAD

**Discussion by franzmondlichtmann** [labwc/labwc#922] [General]
Dear Developers,
How hacky or noob-firendly should labwc become? Will it become super-easy to setup and run?
I mean there are many things that you could add, but otherwise too many (GUI) options could water down the experience.
What are your long-term goals in this matter?
How much do you want to refine your window manager, will it become a full desktop environment?
And thank you very much for all the support/development. Labwc is now my to-go desktop environment for wayland.

**Comment by johanmalm** [comment]
Thanks for writing. Let's me try some quick responses as a starter.

> How hacky or noob-firendly should labwc become?

I much prefer it noob-friendly. I personally find it pretty off-putting when you have to read man pages and wikis to get the basics started.

It might be worth talking some specific examples to align on what we mean by hacky and noob-friendly. My personal view is that labwc is more noob-friendly than hacky, but as you've asked the question, I sense you might not agree :smile:

> Will it become super-easy to setup and run? I mean there are many things that you could add, but otherwise too many (GUI) options could water down the experience.

Well, there is a balance here between what upstream devs (us) do, versus package maintainers (those who package labwc for distributions) versus those who create distributions (or derivative distributions).

We could try to bundle bunch of utilities (e.g. labwc-tweaks and labwc-gtktheme) either in the main repo or in some meta repo, but have so far chosen not to - mostly because we don't want to distract attention from developing the core compositor and getting it to production standard. Also, adding more binary/scripts to the main repo increases commit noise and once they are there, I guess we'd feel a sense of obligation to maintain them. Also, if those binaries/scripts have other dependencies, labwc (as a package) then becomes dependent on those... etc.

> What are your long-term goals in this matter?

Primarily to play the long game an improve the core compositor over time. One big risk to this is burning people out as a result of widening the scope too much.

> How much do you want to refine your window manager, will it become a full desktop environment?

This one is quite simple. We want the compositor to just work really well in its current scope and support official wayland + wlroots protocols. This is pretty hard in itself and what most of us enjoy working on.

Regarding the creation of a Desktop Environment, I suspect that will come, possibly by one of the core devs - or by someone else as a stand alone project. By DE do you mean something like the layer that BunsenLabs have created on top of Openbox and a few clients such as tint2? Or do you mean a much more integrated stack such as mutter + gnome-shell? In the project README, we've laid out some of the key things we stand for in that space.


> And thank you very much for all the support/development. Labwc is now my to-go desktop environment for wayland.

Nice one.

**Comment by franzmondlichtmann** [comment]
@johanmalm 

Yes, I think, you are right becoming a complex desktop environment is probably something another party will do. 

> By DE do you mean something like the layer that BunsenLabs have created on top of Openbox and a few clients such as tint2? Or do you mean a much more integrated stack such as mutter + gnome-shell?

I meant DE in the sense of providing a lot of GUI-Tools for option-management. (I don't really know the terminology)
Personally I think that 99% of options I set, when installing KDE DE for example, are 1 time choices anyway.
I rarely go into the options, mainly to set Icon or GTK/QT Themes, or the font.

Don't really get, why to implement a GUI for so many options one rarely uses anyway. Or making a Desktop Environment with so many bad default features activated, I then want to deactivate (like window snapping, which implementation I don't like in most DEs).
Therefore making a few concious default choices which are attractive to a lot of users without sacrificing on options in a config-file, seems more sane to me.

I know this is a preference, but it seems a lot more easy to set the options via a text file (and more easy to read) then in a options menu (which doesn't provide context with comments like in a text file). 

Maybe making tooltips would be a in-between solution. But tooltips can get only so long (character wise), without being obnoxious.
Another solution would be to provide both, a GUI and a flexible config-file with comments, so you can choose where to set your option. (Config and GUI would set the same options). I know that some people prefer the GUI and its not bad in itself.

I don't like when some option is hidden deep in a nested GUI. :)

Btw. why is programming a window manager fun/necessary? Is it a hobby or a job opportunity? Not that would I question its purpose, I'm just curious.

**Reply by Consolatis** [reply]
> Btw. why is programming a window manager fun/necessary?

For me, I was a Gnome 2 and then Mate user for a long time. When wanting to move to wayland there weren't any low resource stacking options around that I really liked. Labwc ticked the main boxes but had some stuff missing I wanted so I implemented them. And as I am lazy and didn't want to keep my local fork in sync I instead sent PRs to labwc. And somehow I kept being around. Honestly, for me personally labwc is feature complete since like a year or something. But other people have different usage patterns so as long as additional features don't make a mess out of the code base I am fine reviewing them (and potentially implementing them when I feel like it).

Well, and because designing and writing code is fun for me. Even debugging issues can be fun at times.

**Reply by franzmondlichtmann** [reply]
> For me, I was a Gnome 2 and then Mate user for a long time. 

I was a KDE user, and it still has a place in my heart. But I wanted something more simple and modular and wayland based. So I stumpled upon Openbox, but sadly it was for X11, therefore labwc was the way to go. I found it by chance on github, maybe through the arch wiki.
And because it was still maintained and recently updated I tried it. Now I use it on a daily basis on a Laptop (with Nvidia Optimus) more even than my Windows(TM) PC and gave it a star. I definitely can't complain.

**Comment by johanmalm** [comment]
Have you tried https://github.com/labwc/labwc-tweaks

> Btw. why is programming a window manager fun/necessary? Is it a hobby or a job opportunity? Not that would I question its purpose, I'm just curious.

Just a hobby. 

I like the eco-system (wlroots and the wayland protocols) and see beauty in the care that has been taken to create it. That makes it enjoyable to work on.
I think it's also because it's hard and therefore intellectually challenging and interesting.
I also enjoy it because there are no deadlines and no pressure... so we can focus on technical solutions only.

**Reply by franzmondlichtmann** [reply]
> Have you tried https://github.com/labwc/labwc-tweaks

I have not tried it.
But I used nwg-look, which has a lot of options for GTK-Themes, Icon-Themes, Mouse-Cursors, Fonts and so on.
[(https://github.com/nwg-piotr/nwg-look](https://github.com/nwg-piotr/nwg-look)

>I like the eco-system (wlroots and the wayland protocols) and see beauty in the care that has been taken to create it. That makes it enjoyable to work on.
I think it's also because it's hard and therefore intellectually challenging and interesting.
I also enjoy it because there are no deadlines and no pressure... so we can focus on technical solutions only.

Nice, a challenge is always good to stay in the flow.
I also try teaching myself programming. I was told programming previously in a formal IT-apprenticeship, but what the teacher told us was not really helpful. It was far too easy and not a real challenge, and also real beginner stuff, you could learn in a youtube-video within a day. I also hope to learn to a degree, that I can publish my own stuff. But I feel that this may be a long road.

**Reply by Consolatis** [reply]
I usually suggest just writing some software for yourself to solve a specific issue you have. Or a nicer alternative to an application you already use. That will send you on various quests to figure out how to do something specific and when doing that for a while you will get a pretty good generic understanding on how things work. I would recommend starting with some high level language though, for example Python. For the most part, the basic concepts of programming are either the same or at least closely related, regardless of the language you end up using. Low level languages like `C` do have quite a learning curve though. That can be drastically reduced if some generic concepts are already known.

**Reply by franzmondlichtmann** [reply]
Thank you for the tips. For now its python. Can you read minds?
I will progress on my quest, thats the only way to go. :)

**Reply by johanmalm** [reply]
> > Have you tried https://github.com/labwc/labwc-tweaks
> 
> I have not tried it. But I used nwg-look, which has a lot of options for GTK-Themes, Icon-Themes, Mouse-Cursors, Fonts and so on. [(https://github.com/nwg-piotr/nwg-look](https://github.com/nwg-piotr/nwg-look)
> 

`labwc-tweaks` is a GUI for configuring `labwc` (and a small amount of `gsettings`).
`nwg-look` is `gsettings` only.

**Comment by johanmalm** [comment]
Here is a challenge is you want to get stuck into something that's helpful to `labwc`... and it written in python

https://github.com/labwc/labwc-gtktheme

It does a reasonable job today, but it's not perfect. The README contains a couple of challenges.

**Reply by johanmalm** [reply]
I'd be good to port it to C too, so that we can use it in `labwc-tweaks`

**Reply by franzmondlichtmann** [reply]
I will look into it, maybe I can help.
